pac4j-saml-config-db/src/main/java/org/pac4j/saml/dbclient/DbLoadedSamlClients.java

```diff
@@ -0,0 +1,197 @@
+import org.pac4j.core.client.Clients;
+import org.pac4j.core.context.WebContext;
+import org.pac4j.core.util.CommonHelper;
+import org.pac4j.saml.client.SAML2Client;
+import org.pac4j.saml.dbclient.dao.api.SamlClientDao;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * An extension to base {@link Clients}, which works also with clients whose definition is stored in a database. 
+ *
+ * @author jkacer
+ * @since 1.9.0
+ */
+public class DbLoadedSamlClients extends Clients {
+
+    /** SLF4J logger. */
+    private final Logger logger = LoggerFactory.getLogger(DbLoadedSamlClients.class);
+    
+    /** DAO reading definitions of SAML clients from a database. */
+    private SamlClientDao samlClientDao;
+
+    /** Clients set explicitly using a setter. */
+    private List<Client> explicitClients;
+    
+    // ------------------------------------------------------------------------------------------------------------------------------------
+
+    
+	/**
+	 * Creates a new client group with no explicit clients.
+	 */
+	public DbLoadedSamlClients() {
+		super();
+	}
+
+
+	/**
+	 * Creates a new client group with some explicit clients.
+	 * 
+	 * @param clients
+	 *            The clients.
+	 */
+	public DbLoadedSamlClients(Client... clients) {
+		super(clients);
+	}
+
+
+	/**
+	 * Creates a new client group with a single explicit client.
+	 * 
+	 * @param client
+	 *            The client.
+	 */
+	public DbLoadedSamlClients(Client client) {
+		super(client);
+	}
+
+
+	/**
+	 * Creates a new client group with some explicit clients.
+	 * 
+	 * @param clients
+	 *            The clients.
+	 */
+	public DbLoadedSamlClients(List<Client> clients) {
+		super(clients);
+	}
+
+
+	/**
+	 * Creates a new client group with some explicit clients and a callback URL.
+	 * 
+	 * @param callbackUrl
+	 *            The callback URL.
+	 * @param clients
+	 *            The clients.
+	 */
+	public DbLoadedSamlClients(String callbackUrl, Client... clients) {
+		super(callbackUrl, clients);
+	}
+
+
+	/**
+	 * Creates a new client group with some explicit clients and a callback URL.
+	 * 
+	 * @param callbackUrl
+	 *            The callback URL.
+	 * @param client
+	 *            The clients.
+	 */
+	public DbLoadedSamlClients(String callbackUrl, Client client) {
+		super(callbackUrl, client);
+	}
+
+
+	/**
+	 * Creates a new client group with some explicit clients and a callback URL.
+	 *
+	 * @param callbackUrl
+	 *            The callback URL.
+	 * @param clients
+	 *            The clients.
+	 */
+	public DbLoadedSamlClients(String callbackUrl, List<Client> clients) {
+		super(callbackUrl, clients);
+	}
+
+
+	/**
+	 * Initializes all clients by loading them from the database.
+	 * 
+	 * @see org.pac4j.core.util.InitializableWebObject#internalInit(org.pac4j.core.context.WebContext)
+	 */
+	@Override
+	protected void internalInit(final WebContext context) {
+        // Remove all clients first if re-initialized

```

Comment:
I definitely think we should get rid of this class. The initialization of the clients from database should not happen in this class.

Let me be more specific:

- The `internalInit` method of `Clients` is called every time you try to find a client: it only computes the callback url of the client from the callback url of the `Clients`
- Every time you call a method of the client: `redirect`, `getCredentials` or `getUserProfile`,  the `internalInit` method of the client is called: this is where the initialization of the configuration from DB should happen.

The flow I see:
In the `SAML2Client`, the `internalInit(context)` would call `init(clientName, context)` from the `AbstractSAML2Configuration`. So maybe I misleaded you, but the `AbstractSAML2Configuration` should not be a `InitializableWebObject`, it should have a `init(String clientName, WebContext context)` method.
The default `SAML2Configuration` should have an empty `init` method while the `DatabaseSAML2Configuration` should use the DTO to load the configuration associated with the client defined by the `clientName` parameter.

Does it make sense?